File: lc/linkedlist/java/IntersectionOfTwoLists.java
public class IntersectionOfTwoLists {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lengthA = getListLength(headA), lengthB = getListLength(headB);

        if (lengthA >= lengthB) {
            while(lengthA > lengthB) {
                lengthA--;
                headA = headA.next;
            }
        } else {
            while(lengthB > lengthA) {
                lengthB--;
                headB = headB.next;
            }
        }

        while(headA != null && headB != null) {
            if(headA == headB) return headA;
            headA = headA.next;
            headB = headB.next;
        }

        return null;
    }

    private int getListLength(ListNode head) {
        int length = 0;

        while(head != null) {
            length++;
            head = head.next;
        }

        return length;
    }
}

File: lc/linkedlist/ruby/intersection_of_two_lists.rb
def intersection_node(head_a, head_b)
    length_a = list_length head_a
    length_b = list_length head_b

    if length_a >= length_b
        while length_a > length_b
            head_a = head_a.next
            length_a -= 1
        end
    else
        while length_a < length_b
            head_b = head_b.next
            length_b -= 1
        end
    end

    while head_a && head_b
        return head_a if head_a == head_b
        head_a = head_a.next
        head_b = head_b.next
    end
end

private
def list_length(head)
    length = 0

    while head
        head = head.next
        length += 1
    end

    length
end